### PR TITLE
hardening/736_set_replication_factor_and_partitions_in_kafka_agent

### PR DESCRIPTION
**Pull request by pcoello25**
* Fixes issue https://github.com/telefonicaid/fiware-cygnus/issues/736
* 100% unit test passed: 
```
Results :
Tests run: 88, Failures: 0, Errors: 0, Skipped: 0
```
* (unofficial) e2e tests passed:

Parameters added in `kafka_agent.conf`:
```
cygnusagent.sinks.kafka-sink.partitions = 5
cygnusagent.sinks.kafka-sink.replication_factor = 1
```
Creating a topic `roomtest_room` with `partitions` = 5 and `replication_factor` = 1
```
time=2016-02-10T08:47:30.593CET | lvl=INFO | trans=1455090415-940-0000000000 | srv=fiservnoti | subsrv=fiservpathnoti | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionKafkaSink[266] : [kafka-sink] Creating topic at OrionKafkaSink. Topic: roomtest_room , partitions: 5 , replication factor: 1
time=2016-02-10T08:47:30.834CET | lvl=INFO | trans=1455090415-940-0000000000 | srv=fiservnoti | subsrv=fiservpathnoti | function=info | comp=Cygnus | msg=kafka.utils.Logging$class[68] : Topic creation {"version":1,"partitions":{"4":[0],"1":[0],"0":[0],"2":[0],"3":[0]}}
time=2016-02-10T08:47:30.841CET | lvl=INFO | trans=1455090415-940-0000000000 | srv=fiservnoti | subsrv=fiservpathnoti | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionKafkaSink[272] : [kafka-sink] Persisting data at OrionKafkaSink. Topic (roomtest_room), Data ({"headers":[{"fiware-service":"fiservnoti"},{"fiware-servicePath":"fiservpathnoti"},{"timestamp":1455090420186}],"body":{"attributes":[{"name":"temperature","type":"float","value":"1.3"},{"name":"humidity","type":"float","value":"42.3"}],"type":"Room","isPattern":"false","id":"RoomTest"}})
```
Describing kafka `topic`:
```
bin/kafka-topics.sh --describe --zookeeper localhost:2181 --topic roomtest_room
Topic:roomtest_room	PartitionCount:5	ReplicationFactor:1	Configs:
	Topic: roomtest_room	Partition: 0	Leader: 0	Replicas: 0	Isr: 0
	Topic: roomtest_room	Partition: 1	Leader: 0	Replicas: 0	Isr: 0
	Topic: roomtest_room	Partition: 2	Leader: 0	Replicas: 0	Isr: 0
	Topic: roomtest_room	Partition: 3	Leader: 0	Replicas: 0	Isr: 0
	Topic: roomtest_room	Partition: 4	Leader: 0	Replicas: 0	Isr: 0
```
* Assignee: @frbattid

### CODE REVIEW

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 1]
```diff
@@ -258,7 +279,9 @@ private void persistAggregation(KafkaAggregator aggregator) throws Exception {
         if (!topicAPI.topicExists(zookeeperClient, topicName)) {
             LOGGER.info("[" + this.getName() + "] Creating topic " + topicName
                     + " at OrionKafkaSink");
-            topicAPI.createTopic(zookeeperClient, topicName, new Properties());
+            LOGGER.info("[" + this.getName() + "] Partitions: " + partitions + ", "
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: I would add the partitions number and the replication factor to the previous logging trace. Something like "Creating topic at OrionKafkaSink. Topic name: <name>, partitions: <partitiions>, replication factor: <replication_factor>".

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -3,3 +3,4 @@
 - [HARDENING] Add /v1 to Management Interface API routes (#739)
 - [BUG] Fix wrong spec date regarding release/0.12.0 (#750)
 - [HARDENING] Add check for null content-type in notifications (#743) 
+- [HARDENING] Add check for partitions and replication_factor in OrionKafkaSink from agent (#736)
```
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: Add support for configurable number of partitions and replication factor in OrionKafkaSink (#736)

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 3]
```diff
@@ -48,6 +48,8 @@
     private String zookeeperEndpoint;
     private TopicAPI topicAPI;
     private ZkClient zookeeperClient;
+    private int topicPartitions;
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: It is not necessary to use these attributes.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 4]
```diff
@@ -173,6 +180,8 @@ void persistBatch(Batch batch) throws Exception {
         protected String entity;
         protected String attribute;
         protected String topic;
+        protected int partitions;
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Unnecessary

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 5]
```diff
@@ -190,6 +199,14 @@ public String getServicePath() {
             return servicePath;
         } // servicePath
         
+        public int getPartitions() {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Unnecessary

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 6]
```diff
@@ -190,6 +199,14 @@ public String getServicePath() {
...
+            return partitions;
+        } // partitions
+        
+        public int getReplicationFactor() {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Unnecessary

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 7]
```diff
@@ -200,6 +217,8 @@ public void initialize(CygnusEvent cygnusEvent) throws Exception {
             entity = cygnusEvent.getEntity();
             attribute = cygnusEvent.getAttribute();
             topic = buildTopicName();
+            partitions = topicPartitions;
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Unnecessary

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java [hunk 8]
```diff
@@ -250,6 +269,8 @@ public void aggregate(CygnusEvent cygnusEvent) throws Exception {
     private void persistAggregation(KafkaAggregator aggregator) throws Exception {
         String aggregation = aggregator.getAggregation();
         String destination = aggregator.getTopic();
+        int partitions = aggregator.getPartitions();
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Unnecessary

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 1 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: CHANGES_NEXT_RELEASE (hunk 2 above)
Comment (pcoello25) [inline CHANGES_NEXT_RELEASE]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 3 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 4 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 5 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 6 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 7 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java (hunk 8 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionKafkaSink.java]: Fixed in 388a53d

**Comment by frbattid** [comment]
LGTM

**Pull request merged**